lib/Query_Processor.cpp

```diff
@@ -1993,30 +2053,23 @@ Query_Processor_Output * Query_Processor::process_mysql_query(MySQL_Session *ses
 	if (qr == NULL || qr->apply == false) {
 		// now it is time to check mysql_query_rules_fast_routing
 		// it is only check if "apply" is not true
-		if (_thr___rules_fast_routing___keys_values) {
-			char keybuf[256];
-			char * keybuf_ptr = keybuf;
-			const char * u = sess->client_myds->myconn->userinfo->username;
-			const char * s = sess->client_myds->myconn->userinfo->schemaname;
-			size_t keylen = strlen(u)+strlen(rand_del)+strlen(s)+30; // 30 is a big number
-			if (keylen > 250) {
-				keybuf_ptr = (char *)malloc(keylen);
-			}
-			sprintf(keybuf_ptr,"%s%s%s---%d", u, rand_del, s, flagIN);
-			khiter_t k = kh_get(khStrInt, _thr_SQP_rules_fast_routing, keybuf_ptr);
-			if (k == kh_end(_thr_SQP_rules_fast_routing)) {
-				sprintf(keybuf_ptr,"%s%s---%d", rand_del, s, flagIN);
-				khiter_t k2 = kh_get(khStrInt, _thr_SQP_rules_fast_routing, keybuf_ptr);
-				if (k2 == kh_end(_thr_SQP_rules_fast_routing)) {
-				} else {
-					ret->destination_hostgroup = kh_val(_thr_SQP_rules_fast_routing,k2);
-				}
-			} else {
-				ret->destination_hostgroup = kh_val(_thr_SQP_rules_fast_routing,k);
-			}
-			if (keylen > 250) {
-				free(keybuf_ptr);
-			}
+		const char * u = sess->client_myds->myconn->userinfo->username;
+		const char * s = sess->client_myds->myconn->userinfo->schemaname;
+
+		int dst_hg = -1;
+
+		if (_thr_SQP_rules_fast_routing != nullptr) {
+			proxy_debug(PROXY_DEBUG_MYSQL_QUERY_PROCESSOR, 7, "Searching thread-local 'rules_fast_routing' hashmap with: user='%s', schema='%s', and flagIN='%d'\n", u, s, flagIN);
+			dst_hg = search_rules_fast_routing_dest_hg(_thr_SQP_rules_fast_routing, u, s, flagIN);
+		} else if (rules_fast_routing != nullptr) {
+			pthread_rwlock_rdlock(&this->rwlock);
+			proxy_debug(PROXY_DEBUG_MYSQL_QUERY_PROCESSOR, 7, "Searching global 'rules_fast_routing' hashmap with: user='%s', schema='%s', and flagIN='%d'\n", u, s, flagIN);
+			dst_hg = search_rules_fast_routing_dest_hg(rules_fast_routing, u, s, flagIN);
```

Comment:
Micro optimization.
Because `search_rules_fast_routing_dest_hg()` performs a series of `strlen()`, `sprintf()` , optionally `malloc()` and `free()` , the locking should be done inside `search_rules_fast_routing_dest_hg()` .
We can simply pass a `bool` to `search_rules_fast_routing_dest_hg()` to either perform the lock or not.

Reply:
I have added several optimizations, including the suggested ones.